SwiftCompilerSources/Sources/SIL/KeyPath.swift

```diff
@@ -0,0 +1,181 @@
+//===--- KeyPath.swift ----------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import AST
+import SILBridging
+
+//============================================================================//
+// KeyPathPattern
+//============================================================================//
+
+public struct KeyPathPattern {
+  public let bridged: BridgedKeyPathPattern
+
+  public init(bridged: BridgedKeyPathPattern) {
+    self.bridged = bridged
+  }
+}
+
+extension KeyPathPattern {
+  public var components: ComponentArray {
+    ComponentArray(bridged: bridged.components())
+  }
+
+  public var objcString: StringRef {
+    StringRef(bridged: bridged.getObjCString())
+  }
+}
+
+//============================================================================//
+// KeyPathPattern.Component
+//============================================================================//
+
+extension KeyPathPattern {
+  public struct Component {
+    public let bridged: BridgedKeyPathPatternComponent
+
+    public init(bridged: BridgedKeyPathPatternComponent) {
+      self.bridged = bridged
+    }
+  }
+}
+
+extension KeyPathPattern.Component {
+  public struct ComputedPropertyId {
+    public let bridged: BridgedKeyPathPatternComponent.ComputedPropertyId
+
+    public init(bridged: BridgedKeyPathPatternComponent.ComputedPropertyId) {
+      self.bridged = bridged
+    }
+  }
+
+  public enum Kind {
```

Comment:
It is much nicer to make `Component` itself an enum with the cases having the corresponding payloads. E.g.

```
  public enum Component {
    case storedProperty(VarDecl)
    case tupleIndex(Int)
    ...
```

Then of course it cannot have the `bridged` stored property. Instead you'll need to create the corresponding enum case at the place where a Component is constructed, e.g. in `ComponentArray.subscript`, e.g.

```
  public subscript(_ index: Int) -> KeyPathPattern.Component {
    let bridged = bridged.at(index)
    switch bridged.getKind() {
      case .storedProperty: return .storedProperty(bridged.getStoredPropertyDecl().getAs(VarDecl.self))
      ...
    }
  }
```